cirq-core/cirq/transformers/routing/visualize_routed_circuit.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Optional, Tuple, TYPE_CHECKING
+from cirq import circuits, ops
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class _SwapPrintGate(ops.Gate):
+    """A gate that displays the string representation of each qubits on the circuit."""
+
+    def __init__(self, qubits: Tuple[Tuple['cirq.Qid', 'cirq.Qid'], ...]) -> None:
+        self.qubits = qubits
+
+    def num_qubits(self):
+        return len(self.qubits)
+
+    def _circuit_diagram_info_(self, args: 'cirq.CircuitDiagramInfoArgs') -> Tuple[str, ...]:
+        return tuple(f'{str(q[1])}' for q in self.qubits)
+
+
+def routed_circuit_with_mapping(
+    routed_circuit: 'cirq.AbstractCircuit',
+    initial_map: Optional[Dict['cirq.Qid', 'cirq.Qid']] = None,
+) -> 'cirq.AbstractCircuit':
+    """Returns the same circuits with information about the permutation of qubits after each swap.
+
+    Args:
+        routed_circuit: a routed circuit that potentially has inserted swaps tagged with a
+            RoutingSwapTag.
+        initial_map: the initial mapping from logical to physical qubits. If this is not specified
+            then the identity mapping of the qubits in routed_circuit will be used as initial_map.
+    """
+    all_qubits = sorted(routed_circuit.all_qubits())
+    qdict = {q: q for q in all_qubits}
+    if initial_map is None:
+        initial_map = qdict.copy()
+    inverse_map = {v: k for k, v in initial_map.items()}
+
+    ret_circuit = circuits.Circuit(
+        _SwapPrintGate(tuple(zip(qdict.values(), [inverse_map[x] for x in qdict.values()]))).on(
+            *all_qubits
+        )
+    )
+    for m in routed_circuit:
+        # Find the mapping at after this moment
+        for op in m:
+            if ops.RoutingSwapTag() in op.tags:
```

Comment:
What if a non-swap operation has this tag?

Reply:
Minor point, but modeling routing via SWAP gates will affect simulation complexity for simulators that don't support SWAP optimizations (aka `split_untangled_qubits` mode), (currently MPS does not, but it could be added), or for simulating explicitly in `split_untangled_qubits=False` mode.

Aside: to add that capability to any simulator, its `QuantumStateRepresentation` needs to support `reindex` and `kron` methods (supporting `factor` is not required). And then the simulator class itself needs to call the `SimulatorBase` constructor with `split_untangled_qubits=True`. There are examples in `custom_state_simulator_test`.

Reply:
This will only happen if a user manually constructs a circuit (or defined a new routing procedure) that has non-SWAP operations that are tagged with `RoutingSwapTag`, since the routing transformer only adds this tag to SWAPs.
Safe to just ignore if that's the case? I've added an extra check for that.

Reply:
@ammareltigani My main concern here is that users could get bewildering error messages if e.g. the gate involved is a one-qubit gate and this tries to act on its second qubit. Agreed that it's an unlikely case, but I think a guard against it is warranted.

@daxfohl this is a good point, but out of scope for this PR. Could you open an issue to track it?

Reply:
Should such a guard be placed when assigning a non-SWAP operation the tag (I imagine this will require modifying the code in `with_tags()`  in raw_types.py) or for now just raise an exception in this function?

Reply:
@daxfohl I'm not too familiar with how simulators work but Orion's recommendation sounds appropriate to me.

Reply:
It was more of an FYI than any real action required. IDK there's any action to take on it.

Great work on this BTW! I've been out of the loop for a while now but this looks like it'll be a great addition!

Reply:
Thanks @daxfohl ! I'll be presenting this work along with some performance benchmarks comparing with tket on Tuesday. I'll try to also present it in a Cirq Sync meeting in the future.